mycompany-seeq-link-connector-myconnector/src/main/java/com/mycompany/seeq/link/connector/MyConnection.java

```diff
@@ -142,87 +170,95 @@ public void disconnect() {
+        Iterable<DatasourceSimulator.Tag.Value> tagValues = this.datasourceSimulator.getTagValues(
+                parameters.getDataId(),
+                parameters.getStartTime(),
+                parameters.getEndTime(),
+                parameters.getSampleLimit()
+        );
+
+        // Return a stream to iterate through all the samples in the time range.
         //
         // Very important: You must return one sample 'on or earlier' than the requested interval and one sample 'on or
         // later' (if such samples exist). This allows Seeq to interpolate appropriately to the edge of the requested
         // time range.
         //
-        // Streams are important to use here to avoid bringing all of the data into memory to satisfy the
+        // Streams are important to use here to avoid bringing all the data into memory to satisfy the
         // request. The Seeq connector host will automatically "page" the data upload so that we don't hit memory
         // ceilings on large requests. Streams can be created in a variety of ways, such as Guava's
         // Streams.stream(iterable), Java's Stream.of(T... values), or Collection.stream().
         //
         // The code within this function is largely specific to the simulator example. But it should give you an idea of
         // some of the concerns you'll need to attend to.
-        return LongStream.rangeClosed(
-                LongMath.divide(parameters.getStartTime().getTimestamp(), this.samplePeriod.toNanos(),
-                        RoundingMode.FLOOR),
-                LongMath.divide(parameters.getEndTime().getTimestamp(), this.samplePeriod.toNanos(),
-                        RoundingMode.CEILING))
-                .boxed()
-                .map(sampleIndex -> {
-                    TimeInstant key = new TimeInstant(sampleIndex * this.samplePeriod.toNanos());
-                    double value = this.datasourceSimulator.query(Waveform.SINE, key.getTimestamp());
-
-                    return new Sample().key(key).value(value);
-                })
-                .limit(parameters.getSampleLimit())
-                .onClose(() -> {
-                    // If you have any cleanup to do, do it in this onClose block. This is guaranteed to be called if
-                    // iteration is short-circuited for any reason.
-                });
-    }
+        Stream.Builder<Sample> streamBuilder = Stream.builder();
 
-    @Override
-    public Integer getMaxConcurrentRequests() {
-        // This parameter can help control the load that Seeq puts on an external datasource. It is typically
-        // controlled from the configuration file.
-        return this.connectionConfig.getMaxConcurrentRequests();
+        for (DatasourceSimulator.Tag.Value tagValue: tagValues) {
+            streamBuilder.accept(new Sample(tagValue.getTimestamp(), tagValue.getMeasure()));
+        }
+
+        return streamBuilder.build();
```

Comment:
I think it would be more natural to avoid an explicit `Stream.Builder` and instead use something like:

```
return tagValues.map(tagValue -> new Sample(tagValue.getTimestamp(), tagValue.getMeasure()));
```

Reply:
Note: since `tagValues` is an Iterable you would first need to convert it to a stream before `map`ping, using something like `Streams.stream(tagValues.iterator())` (if Guava is available) or `StreamSupport.stream(tagValues.spliterator(), false)`.